### DISCUSSION TITLE
airflow db check TypeError: __init__() got an unexpected keyword argument 'encoding'

### DISCUSSION THREAD

**Discussion by frankie-wy** [apache/airflow#22301] [Q&A]
I build the custom container image, below is my Dockerfile.

FROM apache/airflow:2.2.4-python3.9
USER root
RUN apt-get update \
        && apt-get install -y gcc g++ vim libkrb5-dev build-essential libsasl2-dev \
        && apt-get autoremove -yqq --purge \
        && apt-get clean \
        && rm -rf /var/lib/apt/lists/*
USER airflow
RUN pip install --upgrade pip
RUN pip install apache-airflow-providers-apache-spark --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-3.9.txt
RUN pip install apache-airflow-providers-apache-hdfs --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-3.9.txt"

But i got the error when i run the container

airflow-init_1       | The container is run as root user. For security, consider using a regular user account.
airflow-init_1       | ....................
airflow-init_1       | ERROR! Maximum number of retries (20) reached.
airflow-init_1       |
airflow-init_1       | Last check result:
airflow-init_1       | $ airflow db check
airflow-init_1       | Traceback (most recent call last):
airflow-init_1       |   File "/home/airflow/.local/bin/airflow", line 5, in <module>
airflow-init_1       |     from airflow.__main__ import main
airflow-init_1       |   File "/home/airflow/.local/lib/python3.9/site-packages/airflow/__main__.py", line 28, in <module>
airflow-init_1       |     from airflow.cli import cli_parser
airflow-init_1       |   File "/home/airflow/.local/lib/python3.9/site-packages/airflow/cli/cli_parser.py", line 621, in <module>
airflow-init_1       |     type=argparse.FileType('w', encoding='UTF-8'),
airflow-init_1       | TypeError: __init__() got an unexpected keyword argument 'encoding'
airflow-init_1       |
airflow_airflow-init_1 exited with code 1

**Comment by potiuk** [comment]
This makes no sense. You have to do something that you did not tell us with the container - for example you dynamically install packages or do something else. 

1) The dockerfile you show has mistakes
2) When I correct the mistakes and build the image I cannot reproduce it
3) The error indicates that you have done something completely wrong - the error comes from a completely different command ("connections export") not from "db check" and it looks like your installation is screwed up
4) you are using airlfow image as root (which **could** work but you should not need it - you should use airflow user)
5) you likely did not use the image entrypoint to enter the image (https://airflow.apache.org/docs/docker-stack/entrypoint.html)

If you are using `_PIP_ADDITIONAL_REQUIREMENTS` (which is very likely reason) STOP. this is what screws up running container. (https://airflow.apache.org/docs/docker-stack/entrypoint.html#installing-additional-requirements explains that you shoud not use it for anything else than quick check)

**Reply by frankie-wy** [reply]
@potiuk thanks for the support!!
First, I just run pip install apache-airflow-providers-apache-spark and hdfs packages when building docker image.
Next, run docker-compose up airflow-init command to initial the database. At this time, error messages show up.

**Reply by potiuk** [reply]
CAnnot reproduce. You must have problems with your docker-compose file.

**Reply by frankie-wy** [reply]
Can I use the airflow official docker-compose.yaml to run services?
https://airflow.apache.org/docs/apache-airflow/stable/docker-compose.yaml

**Reply by potiuk** [reply]
1) make sure that you use right docker image, test it manually as described here: https://airflow.apache.org/docs/docker-stack/entrypoint.html#executing-commands (without docker compose)
2) Make sure that you use the right image in your docker-compose
3) if you are not sure how to do either of the two above - please read and learn how docker and docker compose works, there are plenty of tutorials.

If you do not know how docker-compose works, which image it uses, etc. give up the idea of using docker compose before you know exactly what you are doing and switch to regular way of running airlfow in virtual env (https://airflow.apache.org/docs/apache-airflow/stable/start/local.html)

We warned that you need to be a docker-compose expert if you want to use it  https://airflow.apache.org/docs/apache-airflow/stable/start/local.html

**Reply by potiuk** [reply]
And if you want solution that "just works" in production -> as we described in the warning, use Helm Chart and K8S.

**Reply by frankie-wy** [reply]
Attach zip file is my Dockerfile and docker-compose.yaml. I'm pretty sure that i use right docker image.
If I only install apache-airflow-providers-apache-spark, it work!!
[docker_airflow_files.zip](https://github.com/apache/airflow/files/8259842/docker_airflow_files.zip)

But, if i also install hdfs packags, it failed.

**Reply by potiuk** [reply]
Sorry. Can't reproduce.
 I think you need to debug further what the problem is on your side. Experiment and let us know if you find it.

**Comment by potiuk** [comment]
BTW. You have many things wrong:

```
airflow-init_1 | ERROR! Maximum number of retries (20) reached.
```

Like your database is unreachable. 

So I am pretty sure there is more wrong on your system, configuration, setup.

**Comment by potiuk** [comment]
The error lines should not be present if you follow Quick-Start step-by-step.

**Comment by PanShaobo** [comment]
i got the same probolem wih you .
i tried [airflow-2](https://issues.apache.org/jira/browse/AIRFLOW-2).2.4-python3.9/3.8/3.7/3.6 but it didn't work at all
[my stackoverflow](https://stackoverflow.com/questions/71499627/airflow-on-docker-start-get-errortypeerror-init-got-an-unexpected-keywor)

**Comment by potiuk** [comment]
I think I know where your problem is - You are likely not using the "entrypoint" to run airflow. You should always use the original entrypoint https://airflow.apache.org/docs/docker-stack/entrypoint.html and you are lkely overriding it without your own entrypoint which does not initialize airflow environment properly.

**Reply by PanShaobo** [reply]
I tried to add the "entrypoint" at the end of Dockerfile:
`ENTRYPOINT ["/usr/bin/dumb-init", "--", "/entrypoint"]`
`CMD []`

Still error 

My dockerfile and docker-compose file looks like this:
[docker-airflow.zip](https://github.com/apache/airflow/files/8323121/docker-airflow.zip)

reproduce:
1.docker build:`docker build . -f Dockerfile --tag xy/airflow:2.2.4-py3.8`
2.docker compose up:`docker-compose up airflow-init`

**Reply by PanShaobo** [reply]
It seems that there is something wrong with the provider-hdfs. Other providers are fine.

**Comment by PanShaobo** [comment]
I follow [customizing-the-image](https://airflow.apache.org/docs/docker-stack/build.html#customizing-the-image) to build docker container.
But the HDFS extras cause error again

docker build . \
    --pull \
    --build-arg PYTHON_BASE_IMAGE="python:3.8-slim-buster" \  
    --build-arg AIRFLOW_VERSION="${AIRFLOW_VERSION}" \
    --build-arg AIRFLOW_CONSTRAINTS_LOCATION="https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-3.8.txt" \
    --build-arg AIRFLOW_EXTRAS="hdfs" \
    --build-arg INSTALL_MYSQL_CLIENT="false" \
    --build-arg INSTALL_MSSQL_CLIENT="false" \
    --build-arg AIRFLOW_PIP_VERSION="22.0.4" \
    --tag "xy/airflow:0.0.1"

**Comment by potiuk** [comment]
Seems like indeed there is a problem with HDFS provider. I am not sure if we can solve it easily for newer python versions. Please try Python 3.7 and if we see that Python 3.8+ is a problem we will simply make it not compatible with 3.8 if we can't do it otherwise.

**Reply by PanShaobo** [reply]
The problem also exists in 2.2.4-python3.9/3.8/3.7/3.6.
I look forward to this problem being solved in a new airflow version. 
Thanks to the airflow team.

**Comment by jrevillard** [comment]
Same issue with Airflow 2.2.5.. please help...

**Comment by jrevillard** [comment]
For information, the issue is this:

```
argparse==1.4.0
  - snakebite-py3==3.0.5 [requires: argparse]
    - apache-airflow-providers-apache-hdfs==2.2.3 [requires: snakebite-py3]
```

argparse is now in python core and must not be installed. I'm removing it right now and the DB connexion issue is solved... I need to check the hdfs provider now.

**Reply by potiuk** [reply]
Whoa. looks like snakebite-py3 (released last time 2.5 years ago IS the problem). I am afraid our `hdfs` provider is really too old for Python 3.9 and there is no good way to make it works without completely rewriting hdfs provider using different provider. I think we will need to retire it unless someone would like to  rewrite it. I will raise the issue at the devlist.

**Comment by patricktapajos** [comment]
Sorry my english, i´ll try to tell what happened to me. Faced the same problem, but, when removing the lines "RUN pip install...", works!!!. I don´t no what to do to fix it.

**Reply by patricktapajos** [reply]
Hmm, tried what @jrevillard said and, after installing the libs, I uninstalled argparse and it worked fine! Thanks @jrevillard for the tip.

**Comment by PanShaobo** [comment]
Used this at the last of Dockerfile, can work with no error:
`RUN pip uninstall -y argparse`
